src/Agent.Sdk/LogPlugin.cs

```diff
@@ -0,0 +1,534 @@
+                monitorSource.Cancel();
+                await memoryUsageMonitor;
+
+                // job has finished, all log plugins should start their finalize process
+                Dictionary<string, Task> finalizeTasks = new Dictionary<string, Task>();
+                foreach (var plugin in _plugins)
+                {
+                    string typeName = plugin.GetType().FullName;
+                    if (!_shortCircuited[typeName].Task.IsCompleted)
+                    {
+                        _trace.Trace($"Start finalize for plugin '{plugin.FriendlyName}'");
+                        var finalize = plugin.FinalizeAsync(_pluginContexts[typeName]);
+                        finalizeTasks[plugin.FriendlyName] = finalize;
+                    }
+                    else
+                    {
+                        _trace.Trace($"Skip finalize for short circuited plugin '{plugin.FriendlyName}'");
+                    }
+                }
+
+                _trace.Trace($"Wait for all plugins finish finalization.");
+                foreach (var task in finalizeTasks)
+                {
+                    try
+                    {
+                        await task.Value;
+                        _trace.Trace($"Plugin '{task.Key}' finished job finalize.");
+                    }
+                    catch (Exception ex)
+                    {
+                        _trace.Output($"Plugin '{task.Key}' failed with: {ex}");
+                    }
+                }
+
+                _trace.Trace($"All plugins finished finalization.");
+            }
+        }
+
+        public void EnqueueOutput(string output)
+        {
+            if (!string.IsNullOrEmpty(output))
+            {
+                foreach (var plugin in _plugins)
+                {
+                    string typeName = plugin.GetType().FullName;
+                    if (!_shortCircuited[typeName].Task.IsCompleted)
+                    {
+                        _outputQueue[typeName].Enqueue(output);
+                    }
+                }
+            }
+        }
+
+        public void Finish()
+        {
+            _jobFinished.TrySetResult(0);
+        }
+
+        private async Task StartMemoryUsageMonitor(CancellationToken token)
+        {
+            Dictionary<string, Int32> flag = new Dictionary<string, int>();
+            foreach (var queue in _outputQueue)
+            {
+                flag[queue.Key] = 0;
+            }
+
+            _trace.Trace($"Start output buffer monitor.");
+            while (!token.IsCancellationRequested)
+            {
+                foreach (var queue in _outputQueue)
+                {
+                    if (token.IsCancellationRequested)
+                    {
+                        break;
+                    }
+
+                    if (queue.Value.Count > _shortCircuitThreshold)
+                    {
+                        _trace.Trace($"Plugin '{queue.Key}' has too many buffered outputs.");
```

Comment:
I'd do `string pluginName = queue.Key` and then replace all subsequent references to use `pluginName`.

Reply:
good idea.